### DISCUSSION TITLE
Application Insights in Shared Uno Project?

### DISCUSSION THREAD

**Discussion by rizaanlakay** [unoplatform/uno#8604] [General]
Has anyone used Application Insights in an Uno Project? I am trying to use the Logging and Telemetry in a cross-platform way. Here is what I have tried:

Using App Center Analytics and Crashes for IOS, ANDROID and UWP, then having App Center automatically export it to Application Insights. That works fine.

The problem I am having is trying to get Application Insights working on the WASM project.
The WASM project just stops on the following code is run:

            var configuration = new TelemetryConfiguration
            {
                ConnectionString = _connectionString
            };
            
            _telemetryClient = new TelemetryClient(configuration);

Any ideas I could try?

**Comment by jeromelaban** [comment]
AppCenter should be working for Xamarin heads properly. Wasm is not supported, as there's no available support for it from Microsoft. You could try opening an issue on their repository: https://github.com/Microsoft/appcenter.

For app insights, you can find [an example here](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/main/Uno/NuGetPackageExplorer.Wasm/WasmScripts/azure-app-insights.js) in Nuget Package Explorer.

**Reply by rizaanlakay** [reply]
Hi @jeromelaban thanks for replying. I know App Center does not support WASM. My issue is not making application insights working with App Center in WASM. My Issue is trying to make the Application Insights SDK work in WASM, with which I could use the custom events, logging etc. They have Microsoft.ApplicationInsights nuget packages for .net core, services, and a whole bunch of others. When I tried to use the library in a WASM project. It freezes the project and doesnt show up any exceptions. Above is the code snippet I had in the constructor of a View to test it with.

**Reply by jeromelaban** [reply]
Indeed, the original SDK does not work with Wasm. See the link I provided above for a workaround which uses the javascript API. You can still open an issue in the AppInsights side to ask for webassembly support.

**Reply by rizaanlakay** [reply]
Thank you so much, I will look at the workaround now. Much appreciated.

**Reply by rizaanlakay** [reply]
@jeromelaban I'm trying to find resources to use the link you sent me. Would this be the way to call the js from WASM? [https://www.andrewhoefling.com/Blog/Post/uno-platform-wasm-invoke-javascript](https://www.andrewhoefling.com/Blog/Post/uno-platform-wasm-invoke-javascript)

Then also how to call the methods on the application insights JS SDK.

**Reply by jeromelaban** [reply]
This would be the most up to date documentation: https://platform.uno/docs/articles/interop/wasm-javascript-1.html

**Reply by xperiandri** [reply]
https://github.com/microsoft/ApplicationInsights-dotnet/issues/2143

**Reply by xperiandri** [reply]
https://feedback.azure.com/d365community/idea/4f959530-1425-ec11-b6e6-000d3a4f09d0

**Comment by xperiandri** [comment]
I thought about using AppCenter SDK for UWP and Mobile, and for other platforms including WASM using ApplicationInsights.
Then set up auto-export of AppCenter data to blob and create an Azure Function that will import that data into ApplicationInstights.
But I have not implemented that yet.

**Comment by rizaanlakay** [comment]
For anyone who is interested in getting Application Insights working in WASM. I've managed to get it working. Sure its probably not the best, but it works really well.

```js
var appInsights;

define([`https://js.monitor.azure.com/scripts/b/ai.2.min.js`], m => {
    try {
        var snippet = {
            config: {
                instrumentationKey: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
            }
        };
        var init = new m.ApplicationInsights(snippet);;
        appInsights = init.loadAppInsights();
        console.log("Insights loaded.");
    }
    catch (error) {
        console.log("Loading failed: " + error);
    }
});

function CallPageLoad() {
    console.log("Insights intialized.");
    try {
        appInsights.trackEvent({ name: 'WASM Home Page Loaded.' });
        appInsights.trackPageView({ name: "WASM Home Page" });
        console.log("WASM Insights called.");
    }
    catch (err) {
        console.log(err);
    }
}
```

Then from code behind in UNO:

```csharp
#if __WASM__
                WebAssemblyRuntime.InvokeJS("CallPageLoad();");
#endif
```

**Comment by assassin316** [comment]
Is this still the way to use Application Insights in Uno? Or is there an alternative for app insights?
